changes/2020-06-17_encrypt-max-plaintext-length-input/change.md

```diff
@@ -0,0 +1,182 @@
+
+### Known-Length Plaintext
+
+Any plaintext input to the Encrypt where the total length of that plaintext
+can be immediately determined.
+For example, if a user supplies plaintext as a string or byte array to Encrypt,
+that is a known-length plaintext.
+
+### Unknown-Length Plaintext
+
+Any plaintext input to the Encrypt operation where the total length of that plaintext
+cannot be immediately determined.
+For example, if a user supplies plaintext as a stream to Encrypt,
+that is a unknown-length plaintext.
+
+## Summary
+
+When performing the encrypt operation on an unknown length plaintext,
+users MUST be able to specify some input that bounds the length of the plaintext to be encrypted.
+
+This input already exists in the specification as [plaintext length](../../client-apis/encrypt.md#plaintext-length)
+however its effect on the Encrypt operation's behavior is unspecified.
+
+This change renames this input in the specification
+to be a more accurate reflection of its behavior,
+and specifies the correct behavior during the Encrypt operation.
+
+This change also forbids exposing any API which only performs Encrypt on known-length plaintexts
+from allowing users to specify this input.
+
+## Out of Scope
+
+- How different implementations may allow a user to specify this input for the Encrypt operation
+  is out of scope.
+- Significantly changing the shape of any of our Encrypt API implementations is out of scope.
+
+## Motivation
+
+All implementations allow users to specify a plaintext length when calling Encrypt with a
+unknown-length plaintext.
+However, its behavior in the Encrypt operation is not clearly specified.
+
+Additionally, some implementations also allow a plaintext length to be specified on Encrypt with a
+known-length plaintext.
+The behavior of Encrypt in this case is not consistent between implementations.
+
+The purpose of this change is to document the correct behavior of this input and
+rename this input to clarify its intent.
+
+This change renames this input within the spec because
+the intention behind the value of this input is that
+it is a bound on the plaintext length,
+not the exact value.
+This is what the Encrypt operation enforces when this value is set,
+and this is how CMMs should interpret this value in GetEncryptionMaterials if it is set.
+Implementations already treat this input with this intent in mind,
+so we should describe it in the specification as such.
+
+This change specifies that the Encrypt operation MUST pass the real length of the plaintext
+to the CMM if known instead of a user supplied bound because the former is more accurate.
+There is no case where a user would benefit from passing
+a less accurate `max plaintext length` to the CMM.
+If there were such a case,
+then such a CMM is not correctly using the intent behind `maxPlaintextLength`.
+It is better to restrict flexibilty here so that users have less opportunity
+to depend on behaviors from CMMs that are misusing this value.
```

Comment:
To be clear: Using this behavior to try and flush the Caching CMM is not a good idea.
Say, trying to encrypt 0 bytes with a plaintext length of 2**66 to ensure
that the next encrypt gets a new data key.

Reply:
That is a more concrete example of what I'm trying to avoid. Users should not be using this value to be triggering weird behaviors in the CMM. Maybe framing as "the CMM is doing a bad thing" isn't the correct way to frame it though, and instead say "users shouldn't expect setting this value to change the CMM's behavior".